src/devices/src/virtio/vsock/device.rs

```diff
@@ -45,54 +52,155 @@ use super::{defs, defs::uapi, EpollConfig};
+    // a VirtioDevice::activate call into an EventHandler read event which allows the other events
+    // (queue and backend related) to be registered post virtio device activation. That's
+    // mostly something we wanted to happen for the backend events, to prevent (potentially)
+    // continuous triggers from happening before the device gets activated.
+    pub(crate) activate_evt: EventFd,
     device_activated: bool,
 }
 
+// TODO: Detect / handle queue deadlock:
+// 1. If the driver halts RX queue processing, we'll need to notify `self.backend`, so that it
+//    can unregister any EPOLLIN listeners, since otherwise it will keep spinning, unable to consume
+//    its EPOLLIN events.
+
 impl<B> Vsock<B>
 where
     B: VsockBackend,
 {
     /// Create a new virtio-vsock device with the given VM CID and vsock backend.
-    pub fn new(cid: u64, epoll_config: EpollConfig, backend: B) -> super::Result<Vsock<B>> {
-        let mut queue_evts = Vec::new();
-        for _ in defs::QUEUE_SIZES {
-            queue_evts.push(EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?);
-        }
-
-        let queues = defs::QUEUE_SIZES
+    pub fn new(cid: u64, mem: GuestMemory, backend: B) -> super::Result<Vsock<B>> {
+        let queues: Vec<VirtQueue> = defs::QUEUE_SIZES
             .iter()
-            .map(|&s| VirtQueue::new(s))
+            .map(|&max_size| VirtQueue::new(max_size))
             .collect();
+        let mut queue_events = Vec::new();
+        for _ in 0..queues.len() {
+            queue_events.push(EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?);
+        }
 
         Ok(Vsock {
             cid,
+            queues,
+            queue_events,
+            mem,
+            backend,
             avail_features: AVAIL_FEATURES,
             acked_features: 0,
-            epoll_config,
-            backend: Some(backend),
             interrupt_status: Arc::new(AtomicUsize::new(0)),
             interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?,
-            queues,
-            queue_evts,
+            activate_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(VsockError::EventFd)?,
             device_activated: false,
         })
     }
+
+    pub fn cid(&self) -> u64 {
+        self.cid
+    }
+
+    /// Signal the guest driver that we've used some virtio buffers that it had previously made
+    /// available.
+    pub fn signal_used_queue(&self) -> result::Result<(), DeviceError> {
+        debug!("vsock: raising IRQ");
+        self.interrupt_status
+            .fetch_or(VIRTIO_MMIO_INT_VRING as usize, Ordering::SeqCst);
+        self.interrupt_evt.write(1).map_err(|e| {
+            error!("Failed to signal used queue: {:?}", e);
+            DeviceError::FailedSignalingUsedQueue(e)
+        })
+    }
+
+    /// Walk the driver-provided RX queue buffers and attempt to fill them up with any data that we
```

Comment:
Add a comment about what the return value represents.

Reply:
done